File: packages/synthdef-compiler/src/rate.ts
/** Calculation rate of a UGen — the string union the compiler API speaks
 *  (the crate accepts these long forms and the SC short forms alike). */
export type Rate = "scalar" | "control" | "audio";

/** Parse SC short forms `ar` / `kr` / `ir` (case-insensitive). */
export function parseRate(s: string): Rate | null {
  switch (s.toLowerCase()) {
    case "ar":
      return "audio";
    case "kr":
      return "control";
    case "ir":
      return "scalar";
    default:
      return null;
  }
}

File: src-tauri/crates/scsynthdef-compiler/src/rate.rs
use serde::{Deserialize, Serialize};

/// Calculation rate of a UGen, matching the SCgf binary encoding.
#[derive(Debug, Clone, Copy, PartialEq, Eq, PartialOrd, Ord, Hash, Serialize, Deserialize)]
#[repr(i8)]
pub enum Rate {
    Scalar = 0,
    Control = 1,
    Audio = 2,
}

impl Rate {
    pub fn as_i8(self) -> i8 {
        self as i8
    }

    /// Parse from string form. Only the SC short forms `ar` / `kr` / `ir`
    /// are accepted; case-insensitive.
    pub fn parse(s: &str) -> Option<Rate> {
        match s.to_ascii_lowercase().as_str() {
            "ar" => Some(Rate::Audio),
            "kr" => Some(Rate::Control),
            "ir" => Some(Rate::Scalar),
            _ => None,
        }
    }
}
